### PR TITLE
Fix: configuration variable that overrides a template variable in env…

### PR DESCRIPTION
**Pull request by TomerHeber**
…0_environment shows drift

Part of #894

### Solution

Since the change is large and complex, this PR is the first part of fixing the issue.
For now I've just added the Overwrites field to the struct.

### CODE REVIEW

File: client/configuration_variable.go [hunk 1]
```diff
@@ -39,22 +40,30 @@ func (c *ConfigurationVariableSchema) ResourceDataSliceStructValueWrite(values m
...
+	Id             string                           `json:"id,omitempty"`
+	Name           string                           `json:"name"`
+	Description    string                           `json:"description,omitempty"`
+	Type           *ConfigurationVariableType       `json:"type,omitempty" tfschema:",omitempty"`
+	Schema         *ConfigurationVariableSchema     `json:"schema,omitempty"`
+	ToDelete       *bool                            `json:"toDelete,omitempty"`
+	IsReadOnly     *bool                            `json:"isReadonly,omitempty"`
+	IsRequired     *bool                            `json:"isRequired,omitempty"`
+	Regex          string                           `json:"regex,omitempty"`
+	Overwrites     *ConfigurationVariableOverwrites `json:"overwrites,omitempty"` // Is removed when marhseling to a JSON.
```
Comment (TomerHeber) [inline client/configuration_variable.go:66]: This struct is also sent in a POST requests.
Overwrites field should **not** be sent in a request. (We are only intrested of this value in a response).
To solve this issue without making too many code changes, I have decided to write a custom marshal.

File: client/configuration_variable.go [hunk 2]
```diff
@@ -77,6 +86,17 @@ type ConfigurationVariableUpdateParams struct {
 	Id           string
 }
 
+func (v ConfigurationVariable) MarshalJSON() ([]byte, error) {
+	v.Overwrites = nil
+
+	// This is done to prevent an infinite loop.
+	type ConfigurationVariableDummy ConfigurationVariable
```
Comment (TomerHeber) [inline client/configuration_variable.go:93]: without doing the following an infinite loop will happen, because inside marshal we are calling marshal again on the same object.

File: client/configuration_variable.go [hunk 3]
```diff
@@ -77,6 +86,17 @@ type ConfigurationVariableUpdateParams struct {
 	Id           string
 }
 
+func (v ConfigurationVariable) MarshalJSON() ([]byte, error) {
+	v.Overwrites = nil
```
Comment (TomerHeber) [inline client/configuration_variable.go:90]: avoid sending Overwrites when marshaling a json to bytes.

File: client/configuration_variable.go [hunk 4]
```diff
@@ -77,6 +86,17 @@ type ConfigurationVariableUpdateParams struct {
...
+
+	// This is done to prevent an infinite loop.
+	type ConfigurationVariableDummy ConfigurationVariable
+
+	dummy := ConfigurationVariableDummy(v)
+
+	return json.Marshal(&dummy)
```
Comment (TomerHeber) [inline client/configuration_variable.go:97]: this would have caused loops...

File: client/configuration_variable_test.go [hunk 5]
```diff
@@ -291,3 +295,37 @@ var _ = Describe("Configuration Variable", func() {
 		})
 	})
 })
+
+func TestConfigurationVariableMarshelling(t *testing.T) {
```
Comment (TomerHeber) [inline client/configuration_variable_test.go:299]: this test was added to verify that the marshal and unmarshal are working as expected.

File: env0/configuration_variable.go [hunk 6]
```diff
@@ -0,0 +1,143 @@
+package env0
```
Comment (TomerHeber) [inline env0/configuration_variable.go:1]: moved these functions to here from env0/resource_environment.go
(These are mostly commom functions).

File: client/configuration_variable_test.go [hunk 7]
```diff
@@ -187,9 +190,10 @@ var _ = Describe("Configuration Variable", func() {
 	})
 
 	Describe("ConfigurationVariableDelete", func() {
+
 		BeforeEach(func() {
 			httpCall = mockHttpClient.EXPECT().Delete("configuration/"+mockConfigurationVariable.Id, nil)
-			apiClient.ConfigurationVariableDelete(mockConfigurationVariable.Id)
+			_ = apiClient.ConfigurationVariableDelete(mockConfigurationVariable.Id)
```
Comment (sagilaufer1992) [inline client/configuration_variable_test.go:196]: why these changes?

**Review by sagilaufer1992 — APPROVED** [review]

File: client/configuration_variable_test.go (hunk 7 above)
Comment (TomerHeber) [reply client/configuration_variable_test.go:196]: using a new linter - fixes a linting issue.
It tells the linter, I know that this can return an error - but I want to ignore it.
